lasagne/layers/conv.py

```diff
@@ -536,3 +606,176 @@ def convolve(self, input, **kwargs):
+
+    b : Theano shared variable or expression
+        Variable or expression representing the biases.
+
+    Notes
+    -----
+    The transposed convolution is implemented as the backward pass of a
+    corresponding non-transposed convolution. It can be thought of as dilating
+    the input (by adding ``stride - 1`` zeros between adjacent input elements),
+    padding it with ``filter_size - 1 - crop`` zeros, and cross-correlating it
+    with the filters. See [1]_ for more background.
+
+    Example
+    -------
+    To transpose an existing convolution, with tied filter weights:
+
+    >>> from lasagne.layers import Conv2DLayer, TransposedConv2DLayer
+    >>> conv = Conv2DLayer((None, 1, 32, 32), 16, 3, stride=2, pad=2)
+    >>> deconv = TransposedConv2DLayer(conv, conv.input_shape[1],
+    ...         conv.filter_size, stride=conv.stride, crop=conv.pad,
+    ...         W=conv.W, flip_filters=not conv.flip_filters)
+
+    References
+    ----------
+    .. [1]: Vincent Dumoulin, Francesco Visin (2016):
+            A guide to convolution arithmetic for deep learning. arXiv.
+            http://arxiv.org/abs/1603.07285,
+            https://github.com/vdumoulin/conv_arithmetic
+    """
+    def __init__(self, incoming, num_filters, filter_size, stride=(1, 1),
+                 crop=0, untie_biases=False,
+                 W=init.GlorotUniform(), b=init.Constant(0.),
+                 nonlinearity=nonlinearities.rectify, flip_filters=False,
+                 **kwargs):
+        super(TransposedConv2DLayer, self).__init__(
+                incoming, num_filters, filter_size, stride, crop, untie_biases,
+                W, b, nonlinearity, flip_filters, n=2, **kwargs)
+        # rename self.pad to self.crop:
+        self.crop = self.pad
+        del self.pad
+
+    def get_W_shape(self):
+        num_input_channels = self.input_shape[1]
+        # first two sizes are swapped compared to a forward convolution
+        return (num_input_channels, self.num_filters) + self.filter_size
+
+    def get_output_shape_for(self, input_shape):
+        # when called from the constructor, self.crop is still called self.pad:
+        crop = getattr(self, 'crop', getattr(self, 'pad', None))
+        crop = crop if isinstance(crop, tuple) else (crop,) * self.n
+        batchsize = input_shape[0]
+        return ((batchsize, self.num_filters) +
+                tuple(conv_input_length(input, filter, stride, p)
+                      for input, filter, stride, p
+                      in zip(input_shape[2:], self.filter_size,
+                             self.stride, crop)))
+
+    def convolve(self, input, **kwargs):
+        border_mode = 'half' if self.crop == 'same' else self.crop
+        op = T.nnet.abstract_conv.AbstractConv2d_gradInputs(
+            imshp=self.output_shape,
+            kshp=self.get_W_shape(),
+            subsample=self.stride, border_mode=border_mode,
+            filter_flip=not self.flip_filters)
```

Comment:
If you see the operation as: "convolution of dilated input", then `self.flip_filters` should tell whether we do a convolution (`self.flip_filters=True`) or correlation (`self.flip_filters=False`). As gradInputs does exactly the opposite (a correlation if the forward pass is a convolution, and a convolution if the forward pass is a correlation), we need to invert it here. The reasoning is about the same as for renaming `pad` to `crop` -- I would like to present the operation as something straightforward, not as the backward pass of a strided convolution (it just happens to use that for its implementation).